group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/topics/CopartitionedTopicsEnforcerTest.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.streams.topics;
+
+import org.apache.kafka.common.requests.StreamsGroupHeartbeatResponse.Status;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Utils;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.OptionalInt;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.function.Function;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class CopartitionedTopicsEnforcerTest {
+
+    private static final LogContext LOG_CONTEXT = new LogContext();
+
+    private static Function<String, OptionalInt> topicPartitionProvider(Map<String, Integer> topicPartitionCounts) {
+        return topic -> {
+            Integer a = topicPartitionCounts.get(topic);
+            return a == null ? OptionalInt.empty() : OptionalInt.of(a);
+        };
+    }
+
+    @Test
+    public void shouldThrowTopicConfigurationExceptionIfNoPartitionsFoundForCoPartitionedTopic() {
+        final String topic = "topic";
+        final Map<String, Integer> topicPartitionCounts = Collections.emptyMap();
+        final CopartitionedTopicsEnforcer enforcer =
+            new CopartitionedTopicsEnforcer(LOG_CONTEXT, topicPartitionProvider(topicPartitionCounts));
+
+        final TopicConfigurationException ex = assertThrows(TopicConfigurationException.class, () ->
+            enforcer.enforce(
+                Set.of(topic),
+                Set.of(),
+                Set.of()
+            ));
+        assertEquals(Status.MISSING_SOURCE_TOPICS, ex.status());
+        assertEquals("Following topics are missing: [topic]", ex.getMessage());
+    }
+
+    @Test
+    public void shouldThrowTopicConfigurationExceptionIfPartitionCountsForCoPartitionedTopicsDontMatch() {
+        final String firstSourceTopic = "first";
+        final String secondSourceTopic = "second";
```

Comment:
nit:
If you specify all this variables in all tests as constants with sensible names, it would reduce a bit the noise within the tests.

Reply:
Interesting that you think this is better. Done